### ISSUE TITLE
`__hash__ = None` in class definition should not emit assignment error 

### ISSUE THREAD

**Issue by starhel** [python/mypy#16954]
**Bug Report**
To suppress hash support in a class, it's recommended to include `__hash__ = None` in the class definition (it's also required by the ruff rule [PLW1641](https://github.com/astral-sh/ruff/issues/6701)). Unfortunately, mypy reports it as an error.
```
error: Incompatible types in assignment (expression has type "None", base class "object" defined the type as "Callable[[object], int]")  [assignment]
```

**To Reproduce**
https://mypy-play.net/?mypy=latest&python=3.12&gist=2f1c3fd7eff039975b0203c916d75002

```python
class A:
    __hash__ = None

    def __init__(self, x):
        self.x = x

a = A(10)
```

**Comment by JukkaL** [comment]
Closing as duplicate of #4266.
